src/p2p.rs

```diff
@@ -124,10 +128,11 @@ impl Connection {
                     );
                     Ok(block)
                 })?;
-                self.blocks_duration
-                    .observe_duration("process", || func(hash, block));
+                result.push((hash, block));
             }
-            Ok(())
+            let x: Vec<_> = result.into_par_iter().map(|(a, b)| func(a, b)).collect(); // TODO restore observe
```

Comment:
should instead start processing as soon as the data arrives, instead of starting when all the data arrives